### ISSUE TITLE
sstables: close index and data files before unlinking

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#4936]
As a follow up to #4909/#4916 as per @espindola's comment:
> As this test shows, we assume that the unlink of the sstables has released the space on disk.
>
> That is not guaranteed to be the case on unix, since the file can still be open. Can we somehow assert that when we unlink a sstable there are no open file descriptors in scylla?

I'd like to do the following:
* Add sstable::close() method.
* Call sstable::close() from `table::on_compaction_completion` before calling `delete_atomically`.
* Add a warning + backtrace in `sstable::unlink` if `_index_file` or `_data_file` are opened.
* Detach _index_file / _data_file in `sstable::~sstable` before closing asynchronously, before deleting via the '_marked_for_deletion` path - to prevent the above warning.
* Deal with other paths that may trigger this warnings (I presume that some sstable unit tests may hit it too)

**Comment by avikivity** [comment]
It's perfectly legitimate for the file to be in use after compaction, if a query is holding it open. This is increased by querier caching (we may want to evict queriers that hold on to files that are being deleted).

**Comment by espindola** [comment]
Avi Kivity <notifications@github.com> writes:

> It's perfectly legitimate for the file to be in use after compaction, if a query is holding it open. This is increased by querier caching (we may want to evict queriers that hold on to files that are being deleted).

Do we delete the file in that case and depend on the unix semantics or
do we delay deleting the file until the query finishes?

My vote would be to mark the file for deletion and only really delete
and clear the system.large_* tables after it is not being used, since
that is when the space is made available.

Cheers,
Rafael

**Comment by raphaelsc** [comment]
@espindola we'll actually only delete a sstable once its shared_ptr's use count reaches zero. Compaction removes the file from fs's dir entry but file descriptor is kept opened while sstable object is alive.

We decided to do it that way because there's some need for atomic deletion of sstables to prevent data resurrection, look: https://github.com/scylladb/scylla/issues/4082

**Comment by avikivity** [comment]
> Avi Kivity <notifications@github.com> writes:
> It's perfectly legitimate for the file to be in use after compaction, if a query is holding it open. This is increased by querier caching (we may want to evict queriers that hold on to files that are being deleted).
> Do we delete the file in that case and depend on the unix semantics or do we delay deleting the file until the query finishes? My vote would be to mark the file for deletion and only really delete and clear the system.large_* tables after it is not being used, since that is when the space is made available. Cheers, Rafael

It's even called sstable::mark_for_deletion() (and I agree we shouldn't rely on the kernel keeping the file alive).

**Comment by avikivity** [comment]
Although that has an advantage: if we crash, then any files that were held up by running queries will be removed.

**Comment by bhalevy** [comment]
> Although that has an advantage: if we crash, then any files that were held up by running queries will be removed.

Also, there is the atomicity requirement.
We could have a version of delete_atomically that goes half way and prepares the pending_delete log and just marks the sstables for deletion.
They could be actually deleted when the sstable object is destructed upon removal of last reference, and eventually we can complete the removal of the `pending_delete` log.  We'll need to keep a back pointer and reference count from the sstables to the `pending_delete` log that references them.  Not sure it's worth it though...

**Comment by bhalevy** [comment]
Closing as `wontfix`.
With 55cf219c970034496fb4f2136535afea76f4d3cd this is not needed.
